Integration of Google reCAPTCHA into DSpace Feedback Form #3101

VictorDuranEscire opened the pull request:
Hi @tdonohue, I like to share this PR with you.

## References
* Fixes #3101 
* Requires DSpace/DSpace#9866

## Description
Add rechaptcha's verification to feedback form.

URI URL: '/info/feedback'

Add rechaptcha verification to the feedback form to avoid mass comment requests.

## Instructions for Reviewers


List of changes in this PR:
* I have added a setting for the klaro plugin for the reCaptha cookie in the comment form.
* I modified the google reCaptcha service to enable or disable the reCaptha cookie according to the settings on the back end.
* I modified the feedback form to add the reCaptha component, I made the change similar to the registration component.
* I added the messages for the cookie section and the feedback form with all available languages.

I recommend testing in an incognito tab of your browser.


Klaro cookies:

![image](https://github.com/user-attachments/assets/60aeb512-3527-4ea0-a80b-3a9d329f5fed)

Error in the feedback form, button disabled when rejecting cookies:

![image](https://github.com/user-attachments/assets/9f4c4b41-a5be-437b-aa47-fe9851b37a98)

Feedback form when accepting cookies settings:

![image](https://github.com/user-attachments/assets/8d9de7cb-63df-49ec-b478-ff9ec48e6701)

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You do not need to complete this checklist prior creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR **passes [ESLint](https://eslint.org/)** validation using `npm run lint`
- [x] My PR **doesn't introduce circular dependencies** (verified via `npm run check-circ-deps`)
- [x] My PR **includes [TypeDoc](https://typedoc.org/) comments** for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR **passes all specs/tests and includes new/updated specs or tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] My PR **aligns with [Accessibility guidelines](https://wiki.lyrasis.org/display/DSDOC8x/Accessibility)** if it makes changes to the user interface.
- [x] My PR **uses i18n (internationalization) keys** instead of hardcoded English text, to allow for translations.
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

kshepherd commented:
@VictorDuranEscire thanks for this contribution and sorry nobody has gotten to review it until now!

While I review changes and test things out, just a few points that we need to address:
1. There are a few conflicts that need resolving, you can do this with a rebase against the latest main branch
2. Most of the commits in this PR are lint fixes or merge commits - while you are rebasing against main, it would be nice to clean these up by squashing in an interactive rebase (and if possible avoiding merges, preferring rebases instad) -- not a deal-breaker, just helps with reviewing and keeps the commit history tidy.

kshepherd commented:
@tdonohue i have restarted all CI jobs and still see "codecov/project" as "Waiting for status to be reported"... is there any way to force this through?

tdonohue commented:
@VictorDuranEscire : When you have a chance, can you rebase this PR & the corresponding backend PR?  Both have new merge conflicts, which make them difficult to test.  Thanks.

ybnd commented on src/app/core/google-recaptcha/google-recaptcha.service.ts:
```diff
@@ -86,8 +88,15 @@ export class GoogleRecaptchaService {
+      }),
+    );
+    observableCombineLatest([registrationVerification$,feedBackVerification$]).pipe(
+      switchMap(([registrationVerification ,feedBackVerification]) => (registrationVerification || feedBackVerification) ? this.loadRecaptchaProperties() : EMPTY),
```
Can you rewrite this to accomodate future additions? For example:

```typescript
observableCombineLatest([
  registrationVerification$,
  feedBackVerification$,
]).pipe(
  switchMap((verifications) => verifications.some(Boolean) ? this.loadRecaptchaProperties() : EMPTY),
).subscribe();
```

Then it would just be a matter of adding one line to the array, much easier to merge etc.

ybnd commented on src/app/core/feedback/feedback-data.service.ts:
```diff
@@ -61,4 +70,33 @@ export class FeedbackDataService extends IdentifiableDataService<Feedback> imple
+   * @param object    The object to create
+   * @param captchaToken ReCaptchaToken if the verification is enabled
+   */
+  public registerFeedback(object: Feedback, captchaToken: string = null): Observable<RemoteData<Feedback>> {
```
This method appears to be unused

ybnd reviewed (changes requested):
I'm having trouble getting this to work: the form keeps saying I need to accept the cookie for it to work, even though you can see the CAPTCHA is already up in the background, which renders it unusable.
![20250313-084950](https://github.com/user-attachments/assets/4a191ad8-345f-43fe-ba4c-ec1cedbb41c9)

This doesn't happen on the registration page
![20250313-085619](https://github.com/user-attachments/assets/09783643-76fa-4689-be5a-3bfa1b6b259a)

_____

I also noticed this PR re-introduces an empty `klaro-configuration.ts` file, can you get rid of it?

kshepherd commented:
I agree with @ybnd 's feedback - i also wonder if some klaro inclusion / older cookie handling causes that spurious "You have not accepted cookies" state

VictorDuranEscire commented on src/app/shared/cookies/browser-orejime.service.ts:
```diff
@@ -133,15 +138,26 @@ export class BrowserOrejimeService extends OrejimeService {
           appsToHideArray.push(CAPTCHA_NAME);
         }
+        if (hideFeedBackVerification) {
+          appsToHideArray.push(CAPTCHA_FEEDBACK_NAME);
```
@ybnd Okay, I will see if it is possible to make this change and I will comment on what was done at the end.

VictorDuranEscire commented:
Hi @tdonohue @ybnd @kshepherd 

I have solved all the integration conflicts, I also notify that I have made some adjustments:
First I have updated the feedback service with the Orejime service to bring it back to its previous functionality.
Second I updated the way of displaying the proposals to make the messages more dynamic by hiding or displaying as appropriate.
I updated the feedback form with code improvements and considering the previous points. I appreciate your recommendations and look forward to another revision.

tdonohue commented:
@VictorDuranEscire : This PR has somehow gotten "corrupted" (or GitHub has gotten confused by this PR).  The PR is showing that 18,888 lines have been changed and 953 files (see the "Files changed" tab in this PR).  Much of the changes look unrelated to your PR.

Could you rebase this PR against the latest `main` an remove any unrelated commits?  If that proves to be too difficult, you could also try recreating this PR (create a new PR with the same changes, verify it no longer shows the 953 files changed, and close this one).

kshepherd commented:
I would also really recommend avoiding merges, or merge reverts --- it may even be worth starting with a "patch" of changes if you don't have a clean branch anymore, or cherry-picking all your commits over to a fresh branch

In any case, I always think PRs end up tidier and easier to work with if they are only ever rebased when it comes to commit management and keeping up to date with upstream, rather than dealing with merges, reverting commits or merges, etc. Give that a try?

tdonohue commented:
@alexklbuckley : backports are decided on a case by case basis, as it's highly dependent on how complex the backport is (e.g. if an improvement in 9.x requires other code that is only in 9.x, then it cannot be backported easily).

In this case it's not possible to say because this PR is outdated and not in a mergeable state.  So, I don't know the answer as to whether this will get into 10.x and whether it'd be possible to backport.

(I do want to note though that once 10.0 is released, the 7.6.x platform will no longer be supported.  We only support the [last three major releases](https://wiki.lyrasis.org/display/DSPACE/DSpace+Software+Support+Policy).)

alexklbuckley commented:
> @alexklbuckley : backports are decided on a case by case basis, as it's highly dependent on how complex the backport is (e.g. if an improvement in 9.x requires other code that is only in 9.x, then it cannot be backported easily).
> 
> In this case it's not possible to say because this PR is outdated and not in a mergeable state. So, I don't know the answer as to whether this will get into 10.x and whether it'd be possible to backport.
> 
> (I do want to note though that once 10.0 is released, the 7.6.x platform will no longer be supported. We only support the [last three major releases](https://wiki.lyrasis.org/display/DSPACE/DSpace+Software+Support+Policy).)

Thanks for that information @tdonohue , that is helpful
